### ISSUE TITLE
Add UnusedLowMemory to PassOptions?

### ISSUE THREAD

**Issue by kripken** [WebAssembly/binaryen#1919]
We have been missing an optimization opportunity regarding load/store offsets. This proposes that we add logic to PassOptions to help enable it - in general we shouldn't add stuff there for individual optimization needs, so this would be an exception.

Background: we want to optimize
```
(i32.load (i32.add (?) (i32.const 10))
````
into
```
(i32.load offset=10 (?))
````
This is not possible in general, since the add will wrap, unlike the load offset. We currently manage to optimize this when called from emscripten because we know that the lower 1K of memory is not actually used. That means that any offset less than 1K can be optimized, since after wrapping it would not be valid anyhow - so we are replacing undefined behavior with a trap.

This has been done by the `--post-emscripten` pass. Both asm2wasm and the LLVM wasm backend run it. **But**, it was not run at the very end. For example, inlining ran after it. As a result, we missed out on a bunch of opportunities. In addition, we miss out by not running it early, since it could allow other optimizations after it (things that depend on the code size change).

On a rather extreme case - a big interpreter-style loop in zlib, where late optimizations open up opportunities for load offsets - this looks like it may be costing us 1-2% of code size.

I'm suggesting that we add an `UnusedLowMemory` flag to PassOptions. It would live alongside optimization options like `IgnoreImplicitTraps`, `DebugInfo`, and others - basically, a global option that passes need to know about. With that, we can move this optimization into a more proper place in the pipeline, and remove it from `--post-emscripten`.

Thoughts? Ideas for a better name?

**Comment by binji** [comment]
I think this is a good optimization, but I do think it would be better if it could handled by the original wasm producer, rather than as a separate pass in Binaryen. I understand that, at least in LLVM, some of the information is lost by the time the wasm is being generated. But generally that information is known (in the form of pointer to memory region plus member offset, etc), and seems like it could be propagated through the system.

**Comment by kripken** [comment]
Yeah, I believe the wasm backend will do this - but it doesn't manage to do it in all cases, due to technical limitations that I don't know much about. In any case, there will be cases that it can't be expected to do, like if binaryen inlines something after llvm.

**Comment by tlively** [comment]
We now have a `--low-memory-unused` option, so closing this.
